vars/pipeline2ATX.groovy

```diff
@@ -183,6 +224,173 @@ def generateJsonReport(build, attributes, constants, executionTestSteps, logFile
     return JsonOutput.prettyPrint(jsonString)
 }
 
+/**
+ *  Calculates the duration metrics and their percentages.
+ *
+ * @param executionTestSteps
+ *      the stages of the pipeline build
+ * @param build
+ *      the pipeline build
+ * @return a map containing the calculated durations in seconds and percentages
+ */
+def calculateTime(executionTestSteps, build) {
+    def currentPhase = 'setup'
+    int setupDuration = 0
+    int executionDuration = 0
+    int teardownDuration = 0
+    int queueDuration = (build.getStartTimeInMillis() - build.getTimeInMillis()) / 1000
+    def errorTime = calculateErrorTime(executionTestSteps)
+    def fromCommitToStartTime = currentBuild.getBuildCauses('jenkins.branch.BranchEventCause').isEmpty() ? null : getTimeFromCommitToStart(build)
+
+    executionTestSteps.each { stage ->
+        def stageName = stage.name
+        // Update the current phase based on the stage name
+        currentPhase = getCurrentPhase(stageName, currentPhase)
+
+        if (currentPhase == 'teardown') {
+            teardownDuration += stage.duration
+        } else if (currentPhase == 'setup') {
+            setupDuration += stage.duration
+        } else if (currentPhase == 'execution') {
+            executionDuration += stage.duration
+        }
+    }
+
+    int totalDuration = queueDuration + setupDuration + executionDuration + teardownDuration
```

Comment:
how does it differs from [getDuration](https://javadoc.jenkins.io/plugin/workflow-support/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapper.html#getDuration())

Reply:
I've tried that initially, however it seems that `build.getDuration()` always returns `0`